core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java

```diff
@@ -0,0 +1,226 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.cypher.export;
+
+import org.neo4j.graphdb.*;
+import org.neo4j.graphdb.schema.ConstraintDefinition;
+import org.neo4j.graphdb.schema.IndexDefinition;
+import org.neo4j.internal.kernel.api.TokenRead;
+import org.neo4j.kernel.api.KernelTransaction;
+import org.neo4j.kernel.impl.coreapi.InternalTransaction;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static apoc.export.cypher.formatter.CypherFormatterUtils.cypherNode;
+import static apoc.util.Util.quote;
+import static org.neo4j.internal.kernel.api.TokenRead.ANY_LABEL;
+
+public class CypherDatabaseSubGraph implements SubGraph
+{
+    private final Transaction transaction;
+
+    private final String query;
+
+    private final List<Label> labels;
+
+    private final List<RelationshipType> relationshipTypes;
+
+    private final Iterable<Node> nodes;
+
+    private final Iterable<Relationship> relationships;
+
+    public CypherDatabaseSubGraph(Transaction transaction, String query )
+    {
+        this.transaction = transaction;
+        this.query = query;
+        this.nodes = StreamSupport.stream(
+                transaction.execute(query).stream().flatMap(
+                                row -> row.values().stream().filter(value -> value instanceof Node))
+                        .map(node -> (Node) node)
+                        .distinct()
+                        .spliterator()
+                , false
+        ).collect( Collectors.toList());
+        this.relationships = StreamSupport.stream(
+                transaction.execute(query).stream().flatMap(
+                                row -> row.values().stream().filter(value -> value instanceof Relationship))
+                        .map(node -> (Relationship) node)
+                        .distinct()
+                        .spliterator()
+                , false
+        ).collect( Collectors.toList());
+        this.labels = StreamSupport.stream( this.nodes.spliterator(), false )
+                .flatMap(node ->  StreamSupport.stream(node.getLabels().spliterator(), false))
+                .distinct()
+                .collect( Collectors.toList());
+        this.relationshipTypes = StreamSupport.stream( this.relationships.spliterator(), false )
+                .map(relationship ->  relationship.getType())
+                .distinct()
+                .collect( Collectors.toList());
```

Comment:
@BennuFire is this the lazy implementation? generally, I see a lot of `Stream...` and `collect`, meaning that we directly materialize the results, so IMO, this implementation does not guarantee any laziness.

@gem-neo4j wdyt?

Reply:
You are fully right. At the beginning I was aiming for a full lazy implementation but I in that case I had to change some parts of the MultiStatementCypherSubGraphExporter impl. At the moment it's not lazy at all but reactive (avoiding the use of the heap for the tx.result and graph object simultaneously) that's already a memory usage improvement. 

I may change the title in order to avoid misleading although it's an ultimate goal.